### PR TITLE
Create timer that automatically retrieves recipe data 

### PR DESCRIPTION
**Pull request by aidanr67**
So far just added some extra classes and modified TimerWidget and TimerListDialog. Also temporarily disabled some timer related features in Instructions.

### CODE REVIEW

**Comment by JulianVolodia** [comment]
If commits are only for one feature you should squash it into only one and describle all work in description. If they contents more features/bugfixes or sth... you should put it into another pullreq.

Fix that PR, please. Somebody will review the code.

**Comment by rocketman768** [comment]
It's all good...we can worry about squashing when this is no longer a work in progress. This is a pretty neat new feature.

**Comment by rocketman768** [comment]
A lot of compiler warnings/errors.

**Comment by rocketman768** [comment]
In file included from /home/philip/myc/brewtarget/brewtarget/src/TimerDialog.h:9:0,
                 from /home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:1:
/home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.h:66:45: warning: non-static data member initializers only available with -std=c++11 or -std=gnu++11
       BoilTime* boilTime = new BoilTime(this);
                                             ^
/home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.h:67:51: warning: non-static data member initializers only available with -std=c++11 or -std=gnu++11
       QStack<int>* timerPositions = new QStack<int>;
                                                   ^
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp: In constructor ‘TimerDialog::TimerDialog(TimerListDialog*, BoilTime*)’:
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:24:1: error: expected identifier before ‘{’ token
 {
 ^
In file included from /home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:1:0:
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.h:70:22: warning: ‘TimerDialog::mainTimer’ will be initialized after [-Wreorder]
     TimerListDialog* mainTimer;
                      ^
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.h:55:14: warning:   ‘QPalette TimerDialog::paletteOld’ [-Wreorder]
     QPalette paletteOld, paletteNew;
              ^
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:12:1: warning:   when initialized here [-Wreorder]
 TimerDialog::TimerDialog(TimerListDialog *parent, BoilTime* bt) :
 ^
In file included from /home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:1:0:
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.h:56:10: warning: ‘TimerDialog::oldColors’ will be initialized after [-Wreorder]
     bool oldColors;
          ^
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.h:50:22: warning:   ‘Ui::TimerDialog* TimerDialog::ui’ [-Wreorder]
     Ui::TimerDialog *ui;
                      ^
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:12:1: warning:   when initialized here [-Wreorder]
 TimerDialog::TimerDialog(TimerListDialog *parent, BoilTime* bt) :
 ^
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp: In member function ‘void TimerDialog::reset()’:
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:210:5: error: ‘mediaPlayer’ was not declared in this scope
     mediaPlayer->stop();
     ^
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp: In member function ‘void TimerDialog::on_stopButton_clicked()’:
/home/philip/myc/brewtarget/brewtarget/src/TimerDialog.cpp:225:5: error: ‘mediaPlayer’ was not declared in this scope
     mediaPlayer->stop();
     ^
[ 60%] Building CXX object src/CMakeFiles/btobjlib.dir/unit.cpp.o
src/CMakeFiles/btobjlib.dir/build.make:2320: recipe for target 'src/CMakeFiles/btobjlib.dir/TimerDialog.cpp.o' failed
make[2]: *** [src/CMakeFiles/btobjlib.dir/TimerDialog.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.cpp:20:0:
/home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.h:66:45: warning: non-static data member initializers only available with -std=c++11 or -std=gnu++11
       BoilTime* boilTime = new BoilTime(this);
                                             ^
/home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.h:67:51: warning: non-static data member initializers only available with -std=c++11 or -std=gnu++11
       QStack<int>* timerPositions = new QStack<int>;
                                                   ^
/home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.cpp: In member function ‘void TimerListDialog::on_loadRecipesButton_clicked()’:
/home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.cpp:174:26: warning: unused variable ‘add’ [-Wunused-variable]
         QAbstractButton *add = mb.addButton(tr("Add"), QMessageBox::NoRole);
                          ^
/home/philip/myc/brewtarget/brewtarget/src/TimerListDialog.cpp:188:25: warning: comparison between ‘enum Hop::Use’ and ‘enum TimerListDialog::on_loadRecipesButton_clicked()::Use’ [-Wenum-compare]
         if (h->use() == boil) {
                         ^
CMakeFiles/Makefile2:249: recipe for target 'src/CMakeFiles/btobjlib.dir/all' failed
make[1]: *** [src/CMakeFiles/btobjlib.dir/all] Error 2
Makefile:147: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by aidanr67** [comment]
Yeah,

Initiation of the "boiltime" object with parent object "Timerlistdialog" is
causing some errors, I'll sort it out in the next couple of days. Should
still compile so you can see what it looks like for now.

I used it on a brew day today and have some modifications to make but its
mostly aesthetic.

File: src/TimerDialog.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+    timeLCD->repaint();
+}
+
+void TimerDialog::reset()
+{
+    time = boilTime->getTime() - (setTimeBox->value() * 60);
+    if (stopped)
+        stopped = false;
+    mediaPlayer->stop();
```
Comment (rocketman768) [inline src/TimerDialog.cpp]: Not wrapped in `#ifndef NO_QTMULTIMEDIA` causes compiler error.

File: src/TimerDialog.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+#ifndef NO_QTMULTIMEDIA
+   mediaPlayer->play();
+#endif
+   stopButton->setEnabled(true);
+}
+
+void TimerDialog::on_stopButton_clicked()
+{
+    //stop button
+    mediaPlayer->stop();
```
Comment (rocketman768) [inline src/TimerDialog.cpp]: This needs to be wrapped in `#ifdef` too

File: src/TimerDialog.cpp (hunk 1 above)
Comment (aidanr67) [inline src/TimerDialog.cpp]: Aah cool, I'll sort that out. I took that from your code.

File: src/TimerDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+#include <QMessageBox>
+
+TimerDialog::TimerDialog(TimerListDialog *parent, BoilTime* bt) :
+    QDialog(parent),
+    mainTimer(parent),
+    paletteOld(),
+    paletteNew(),
+    oldColors(true),
+    ui(new Ui::TimerDialog),
+    boilTime(bt),
```
Comment (rocketman768) [inline src/TimerDialog.cpp]: This causes a compiler error when `NO_QTMULTIMEDIA` is defined due to the trailing comma. Move comma to the beginning of the line, e.g.

    , mediaPlayer(...)
    , playlist(...)

**Comment by aidanr67** [comment]
Okay, compiler warning are fixed and I'm mostly happy with where it's at now. Please have a look and let me know if there's any issues or ways I can improve it.

File: src/TimerDialog.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    timeLCD->repaint();
+}
+
+void TimerDialog::reset()
+{
+    //Resetting doesn't stop timers -- should it?
+    time = boilTime->getTime() - (setTimeBox->value() * 60);
+    if (stopped)
+        stopped = false;
+#ifdef NO_QTMULTIMEDIA
```
Comment (rocketman768) [inline src/TimerDialog.cpp]: `#ifndef`

**Comment by rocketman768** [comment]
I think the functionality here is really good.

The major suggestion I have is to pull all the timers together into one cohesive dialog. 10 dialogs on my IPA recipe really clutters the screen :smiley:. I'd like to see one big dialog with the main clock and stuff up at the top, then a scroll area with a grid layout populated with the little timers...roughly like

layout

**Comment by aidanr67** [comment]
Yeah, I thought about doing it that way. The nice thing about doing it this
way though is that you can hide all the timers and they'll individually pop
up as they go off. What about just hiding them by default? That'll only
show the main timer.

**Comment by aidanr67** [comment]
I definitely agree that 10 timers scattered across the screen is not pretty, maybe I can do more to control their initial layout.

 My feeling is that users won't really want to modify the automatically generated timers, unless they really want to change the alarm sound, so control of the timers is handled by the main timer and there's no need really to view the individual timers. Users may want to add additional (non recipe related) timers, for example adding a coiling coil to the boil, so those will be shown by default as they are now. What about hiding automatically generated timers by default, and maybe pushing a notification to  the OS, something like: "7 timers generated, click show timers to view them.".

I like the visual "NOTICE ME!" of the timers popping up as they go off. If I put them all into a single dialog with scroll it'll have to auto scroll to center a timer that is going off but I don't think that will have the same visual effect as the pop up.

Anyway, thanks for the suggestion, I'll look into it. I'm keen to here more views on this because I feel both ways have there benefits. Maybe we can do a hybrid of sorts.

**Comment by theophae** [comment]
Really cool feature :)
This software is getting better.
I have an idea to avoid those multiple windows:
It could be a good idea to condensate all you've put on each window in one line using a few icons like this
timers

**Comment by aidanr67** [comment]
That's a cool idea, thanks for the input.

I'll play around with it and see what I can come up with.

**Comment by theophae** [comment]
Your welcome ! You did the hard work ;)

**Comment by aidanr67** [comment]
I'm quite happy with how it's looking now.

screenshot from 2015-12-09 10-01-23

**Comment by JakeFry** [comment]
Hi and thanks for this useful feature.
I've found a bug under windows. If I have 30,000g of hops to add in the recipe timer shows that I have to add 29g. If I change it to 30,001g it's 30g as it should so there is some conversion and rounding issue. The other thing is that after closing the Main boil Timer window the there is way to move or close the timers window. You have to drag the main program window from under the timers window to select Tools->Timers and click Hide Timers.

image

**Comment by aidanr67** [comment]
Hey, thanks for the feedback. I'll look into these issues. I haven't tried
on Windows but the escape key should hide the timer window.

**Comment by rocketman768** [comment]
OK. Overall design is shaping up. There are a few UI things to clean up:

1. Broken layout and unecessary QFrame in timerWidget.ui.
2. Size of timerMainDialog is fixed. Reset the min/max size to defaults. You can set a minimum size for timeLCD if you want the dialog to be big.
3. Layout of timerMainDialog is broken.
4. Size of timerWidget is fixed. Reset the min/max size to defaults.
5. Size of TimerListDialog is fixed. Remove the lines in the dialog's constructor under `setWindowTitle()`.
6. Add `scrollArea->setWidgetResizable(true);` somewhere in TimerListDialog's constructor.
7. Add horizontal spacers to one or both sides of the buttons so the button is only as big as the text it contains.

**Comment by aidanr67** [comment]
Cool, will sort all that out.

Is there an easy way to access the weightSystemUnit so that I can convert g
to oz depending on selected system. I looked at using a BtLineEdit but I
don't think it'll work since I have the double and text in the same
lineEdit.

**Comment by aidanr67** [comment]
@JakeFry - My last commit should have resolved your hop amount issue, I couldn't replicate your issue on my side but I removed the type conversion from double to int so there is no rounding issues.

**Comment by JakeFry** [comment]
I can confirm that the hop amount issue is corrected with this commit

**Comment by rocketman768** [comment]
@aidanr67,

About unit conversion, have a look at [Brewtarget::displayAmount()](https://github.com/Brewtarget/brewtarget/blob/master/src/brewtarget.h#L184). It is used to convert amounts in arbitrary units into a localized QString in the units the user wants to see. It's used all over the place in the code if you need some examples.

File: src/TimerMainDialog.cpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    bool timerFound = false;
+    int duplicates = 0;
+    int timersGenerated = 0;
+    bool duplicatesFound = false;
+    QList<Hop*> hops;
+    QString note;
+    hops = recipe->hops();
+    foreach (Hop* h, hops) {
+        if (h->use() == 2) { //2 = Boil addition -- Hop::Use enum
+            note = Brewtarget::displayAmount(h->amount_kg(), "TimerNote", "hop_amount", Units::kilograms) +
```
Comment (rocketman768) [inline src/TimerMainDialog.cpp]: Make sure all user-visible strings are translatable with [QObject::tr()](http://doc.qt.io/qt-5/qobject.html#tr). For example, this one should be

    note = tr("%1 of %2").arg(Brewtarget::displayAmount(...)).arg(h->name());

Comment (aidanr67) [inline src/TimerMainDialog.cpp]: Cool, will do.

File: src/TimerDialog.cpp [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    updateTime();
+}
+
+void TimerDialog::setNote(QString n)
+{
+    //Append if notes exist, new note if not
+    if (noteEdit->text() == "Notes..." || noteEdit->text() == "")
+        noteEdit->setText(n);
+    else
+        noteEdit->setText(noteEdit->text() + " and " + n);
```
Comment (theophae) [inline src/TimerDialog.cpp]: Missing tr()

File: src/TimerDialog.cpp [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+   if( !playlist->addMedia(QUrl::fromLocalFile(s)) )
+      Brewtarget::logW(playlist->errorString());
+   playlist->setCurrentIndex(0);
+#endif
+}
+
+void TimerDialog::on_setTimeBox_valueChanged(int t)
+{
+    if (t*60 > boilTime->getTime()) {
+        QMessageBox::warning(this, "Error", "Addition time cannot be longer than remaining boil time");
```
Comment (theophae) [inline src/TimerDialog.cpp]: Missing tr()

Comment (aidanr67) [inline src/TimerDialog.cpp]: Aah, good spotting. Will sort that out.

**Comment by rocketman768** [comment]
@aidanr67 The last commit I see is 4da454eb. Do you have another commit after that which fixes the missing `tr()` calls?

**Comment by aidanr67** [comment]
Sorry guys, I see where the confusion came in. I removed TimerDialog class from the project and replaced with TimerWidget class. I thought the commit would remove the files but I had to go into Github and remove manually. Sorry @theophae, you where right, I was looking at the code and not the file names.

**Comment by rocketman768** [comment]
@aidanr67 

:+1: 

Let's get this merged! There is a conflict with the current master right now in `TimerWidget.cpp`. After that is fixed, I'll merge.

**Comment by aidanr67** [comment]
@rocketman768 

Awesome, should be good to go now.

**Pull request merged**